US354240: Provide CAF Logging support for Serilog

xbreizh opened the pull request:
https://internal.almoctane.com/ui/entity-navigation?p=131002/6001&entityType=work_item&id=354240

buildmachine-sou-jenkins2 commented:
A developer build has not yet been created for this branch. Click [here](http://sou-jenkins2.swinfra.net:8080/job/Create-New-Build/parambuild/?GITHUB_SERVER=github.com&GITHUB_ORGANISATION=CAFapi&REPO_NAME=caf-logging&BRANCH_NAME=US354240) to go ahead and create the build...

buildmachine-sou-jenkins2 commented:
CI Build Link:
http://sou-jenkins2.swinfra.net:8080/job/CAFapi/job/CAFapi~caf-logging~US354240~CI

xbreizh committed:
Add caf-logging-dotnet

xbreizh committed:
Update .gitignore

xbreizh committed:
Add vs code

xbreizh committed:
add official-build

xbreizh committed:
frank

xbreizh committed:
building dotnet module init

xbreizh committed:
Add phases

xbreizh committed:
test

xbreizh committed:
only library

xbreizh committed:
using profiles

xbreizh committed:
rename module + add build phase

xbreizh committed:
remove skip deploy

xbreizh committed:
deploy false

xbreizh committed:
implement serilog wrapper

xbreizh committed:
cleanup / rename

xbreizh committed:
library only

xbreizh committed:
Add package phase

xbreizh committed:
push attempt

xbreizh committed:
Using caf-parent

xbreizh committed:
Using buildenv-dotnet-image

xbreizh committed:
using fabric8

xbreizh committed:
jdk11

xbreizh committed:
push

xbreizh committed:
test

xbreizh committed:
jdk8

xbreizh committed:
plugin1

xbreizh committed:
rename package

xbreizh committed:
fix sanitize

xbreizh committed:
Fix test and formatting

xbreizh committed:
Add fixes and test class file

xbreizh committed:
Update regex replace

xbreizh committed:
add package folder

xbreizh committed:
update solution + tests

xbreizh commented:
I added [Program.cs](https://github.com/CAFapi/caf-logging/blob/US354240/caf-logging-serilog/src/main/resources/Program.cs) which allows testing of the library.
I fixed the sanitizing process as per other modules (log4j, logback).
I added a new Object, [LogExceptionItem](https://github.com/CAFapi/caf-logging/blob/2c3101b8c0575164fe9111d2c622a82b65e43d8e/caf-logging-serilog/caf-logging.serilog/LogExceptionItem.cs) as a helper to create the json object in case of exception or unsafe message.

For updating the properties, I am using [SimpleClassEnricher](https://github.com/CAFapi/caf-logging/blob/2c3101b8c0575164fe9111d2c622a82b65e43d8e/caf-logging-serilog/caf-logging.serilog/SimpleClassEnricher.cs). This allows filtering and sanitizing as we need.

xbreizh committed:
Update solution

xbreizh committed:
file rename

dermot-hardy commented on .gitignore:
```diff
@@ -10,11 +10,15 @@ Berksfile.lock
+obj/
+bin/
+.vs/
+sepg-dotnet-build-adapter.csproj.user
```
What is this line?

xbreizh replied:
That's a local configuration file from Visual Studio. The name is no longer accurate so it has been removed.

dermot-hardy commented on caf-logging-log4j2/src/test/java/com/github/cafapi/logging/log4j2/LogInjectionTests.java:
```diff
@@ -56,6 +58,11 @@ public void testLogMessageWithNewline() throws Exception
+        System.out.println("log " + log);
+        String sd = "\n";
+        IntStream df = sd.chars();
+        System.out.println("lb " + df.findFirst().getAsInt());
```
Why would adding Serilog support involve making any changes to the Log4J2 tests?

xbreizh replied:
Quick test, I removed it

dermot-hardy commented on caf-logging-serilog/.gitignore:
```diff
@@ -0,0 +1,310 @@
+## Ignore Visual Studio temporary files, build results, and
+## files generated by popular Visual Studio add-ons.
+
```
If we're creating a `.gitignore` file at this level then do we really need the changes to the repository level `.gitignore` file?

xbreizh replied:
Now using vs specific only at that level, and replaced the content with source from [here](https://github.com/github/gitignore/blob/main/VisualStudio.gitignore)

dermot-hardy commented on caf-logging-serilog/official-build.props:
```diff
@@ -0,0 +1 @@
+SEPG_BUILD_ENV_IMAGE=cafapi/buildenv-dotnet:1.0.0
```
Having an `official-build.props` file at this level can't be doing anything.  SEPG will only pay attention to the repository-level one.

xbreizh replied:
Removed

dermot-hardy commented on pom.xml:
```diff
@@ -69,12 +69,15 @@
         <enforceCorrectDependencies>true</enforceCorrectDependencies>
         <logbackVersion>1.2.3</logbackVersion>
+        <maven.install.skip>true</maven.install.skip>
+        <maven.deploy.skip>true</maven.deploy.skip>
```
These changes seem worrying.  Surely this will inherit to all modules and result in nothing being produced from this project?

xbreizh replied:
Removed

dermot-hardy commented on pom.xml:
```diff
@@ -143,5 +146,25 @@
+                    </execution>
+                </executions>
+                <configuration>
+                    <skipNexusStagingDeployMojo>true</skipNexusStagingDeployMojo>
```
Same with this.

xbreizh replied:
Removed

dermot-hardy commented on caf-logging-serilog/src/main/docker/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+RUN wget https://packages.microsoft.com/config/opensuse/15/prod.repo
+RUN mv prod.repo /etc/zypp/repos.d/microsoft-prod.repo
+RUN chown root:root /etc/zypp/repos.d/microsoft-prod.repo
+RUN zypper -n install dotnet-sdk-5.0
```
Why do we have a `Dockerfile`?

xbreizh replied:
I was just testing different things to have the nuget generated and pushed. Removed

dermot-hardy commented on caf-logging-serilog/src/main/docker/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+# limitations under the License.
+#
+
+FROM dockerhub-private.svsartifactory.swinfra.net/cafapi/opensuse-jdk8-maven:1.0.0
```
- No open-source component can ever refer to internal company-private services.  (I know there are a few instances where this rule is broken but they are bugs.)
- Also we don't use `dockerhub-private` even internally anymore (except maybe for WebInspect).

xbreizh replied:
Removed

dermot-hardy reviewed (changes requested):
We'll go through the C# modules together.  I'm not going to start commenting on them here.

xbreizh committed:
review comments

xbreizh committed:
Merge branch 'master' into US354240

dermot-hardy committed:
Use .NET build environment

dermot-hardy committed:
caf-logging-serilog pom changes

dermot-hardy committed:
Added license headers

dermot-hardy committed:
Upgraded license plugin

dermot-hardy committed:
Configured license plugin

dermot-hardy committed:
Restored install for symmetry with deploy

xbreizh committed:
moved from Debug to Release

xbreizh commented on caf-logging-serilog/caf-logging-serilog.sln:
```diff
@@ -11,8 +11,8 @@ Global
 		Release|Any CPU = Release|Any CPU
 	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
-		{540E19A9-A196-4BFC-8F92-7F55608726F0}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
```
https://docs.microsoft.com/en-us/visualstudio/debugger/how-to-set-debug-and-release-configurations?view=vs-2022

xbreizh commented on caf-logging-serilog/caf-logging-serilog.sln:
```diff
@@ -11,8 +11,8 @@ Global
 		Release|Any CPU = Release|Any CPU
 	EndGlobalSection
```
test

xbreizh reviewed (commented):
test

xbreizh committed:
Update namespace

xbreizh committed:
Merge branch 'US354240' of github.com:CAFapi/caf-logging into US354240

xbreizh committed:
update dotnet solution

xbreizh committed:
formatting

xbreizh committed:
Add TODO

xbreizh committed:
prevent exceptions

dermot-hardy committed:
Merge branch 'master' into US354240

xbreizh committed:
review updates

xbreizh committed:
update template + unsecure message mgt

xbreizh committed:
Add missing classes

xbreizh committed:
add functions

xbreizh committed:
Using functions

xbreizh committed:
remove obsolete functions

xbreizh committed:
refactoring

xbreizh committed:
Fix header

xbreizh committed:
Fix + add file sink config

xbreizh committed:
Add sink file

xbreizh commented:
@frankmccarry I added the file configuration and it performs as expected. The only downside is that we have to pass the filepath to the configuration as it's a requirement from WriteTo.File . I tested with a default name, but then the template doesn't get applied onto the file mentioned in .ReadFrom.Configuration(configuration).

Here is a sample of call:

    var config_path = "C:\\Users\\XLamourec\\IdeaProjects\\caf-logging\\ConsoleApp1\\appsettings.json";
    var configuration = new ConfigurationBuilder()
               .AddJsonFile(config_path)
               .Build();

     var levelSwitch = new LoggingLevelSwitch();
     levelSwitch.MinimumLevel = LogEventLevel.Verbose;
     Log.Logger = CafLoggingLoggerConfiguration
         .GetLoggerConfig(levelSwitch, "testlog.txt")
         .ReadFrom.Configuration(configuration)
         .CreateLogger();

xbreizh committed:
Add Readme

xbreizh committed:
remove unused functions

xbreizh committed:
restore log4j logging

xbreizh committed:
review changes

frankmccarry reviewed (approved):
I think the changes meet the remit for the ticket.

dermot-hardy commented on caf-logging-log4j2/src/main/resources/log4j2.xml:
```diff
@@ -20,7 +20,7 @@
         <Console name="STDERR" target="SYSTEM_ERR">
             <PatternLayout
-                pattern="[%d{yyyy-MM-dd HH:mm:ss.SSS}{UTC}Z %pidtid %-5level ${%-12.-12sanitize{ %X{tenantId} }:\-} ${%-4.-4sanitize{ %X{correlationId} }:\-}] %sanitize{ %logger{30} }: %maybeJsonMsgAndEx%n" 
+                pattern="[%d{yyyy-MM-dd HH:mm:ss.SSS}{UTC}Z %pidtid %-5level ${%-12.-12sanitize{ %X{tenantId} }:\-} ${%-4.-4sanitize{ %X{correlationId} }:\-}] %sanitize{ %logger{30} }: %maybeJsonMsgAndEx%n"
```
Checked in 1b704bf for this.  It is clearly not part of this ticket.

dermot-hardy committed:
Merge branch 'master' into US354240

dermot-hardy committed:
Fix the warnings in the build log

dermot-hardy committed:
Removed the duplicate function

dermot-hardy committed:
Corrected line endings in previous revision

dermot-hardy committed:
Fixed the namespaces

Also moved the source files alongside the project per .NET conventions

dermot-hardy committed:
Updated to stay within guideline

dermot-hardy commented on caf-logging-serilog/Sanitizer.cs:
```diff
@@ -0,0 +1,135 @@
+            {
+                if (char.IsControl(c))
+                {
+                    return true;
```
So we're saying that if the message being logged contains a control character that it is safe to log!!

xbreizh replied:
Reversed the logic

dermot-hardy commented on caf-logging-serilog/Sanitizer.cs:
```diff
@@ -0,0 +1,135 @@
+            // http://www.csc.villanova.edu/~tway/resources/ascii-table.html
+            foreach (char c in message)
+            {
+                if (char.IsControl(c))
```
This check doesn't relate to the comment above.  Are we checking if the characters are ASCII, or are we checking if they are control characters?

xbreizh replied:
I updated to match what we do [here](https://github.com/CAFapi/caf-logging/blob/1b704bf34f81f476e6cf928cfb26eb5d0972490d/caf-logging-common/src/main/java/com/github/cafapi/logging/common/LogMessageValidator.java#L49) Note that we consider non-latin characters as unsafe.

dermot-hardy commented on caf-logging-serilog/README.md:
```diff
@@ -0,0 +1,67 @@
+
+## Examples
+
+`[2022-05-10 13:29:45.444Z #217.001 WARN  john-tenant  Cidi] c.m.d.f.f.f.FieldFullTextFixer: "Greek test: κόσμε"`
```
Why does the strings being logged have quotes around them?  And also why are there extra quotes in the strings in the `message` field?

I.e. I tried this: `Log.Error(ex, "Caught an exception");`

I expected: `{"message":"Caught an exception","exception":"..."}`
But I got: `{"message":"\"Caught an exception\"","exception":"..."}`

xbreizh replied:
Fixed

dermot-hardy commented on caf-logging-serilog/README.md:
```diff
@@ -0,0 +1,67 @@
+
+## Pattern
+
+`[{@t:yyyy-MM-dd HH:mm:ss.fffZ} {Tid(ProcessId,ThreadId)} {Log(@l):5} {Sanitize(tenantId, 12, 12)} {Sanitize(correlationId, 4, 4)}] {Sanitize(logger, 30, 30)}: {MaybeJsonMsgAndEx(@m,@x)}\n`
```
- When I tested this I did not see the time coming out in UTC.
- Tenant Id and Correlation Id are not defaulting to a dash if they are not present

xbreizh replied:
Both should be fixed now. I am getting: `[2022-05-13 15:10:53.160Z #272.001 ERROR - -] c.m.d.f.f.f.FieldFullTextFixer: {"message":"{FirstName"}`

dermot-hardy commented on caf-logging-serilog/README.md:
```diff
@@ -0,0 +1,67 @@
+# CAF Logging Serilog
+This project configures Serilog to meet the CAF Logging Standard.
+
+The log level is controllable via value that can be provided on Logger initialization. The default used is `INFO`.
```
Can't we control the log level by setting the `CAF_LOG_LEVEL` environment variable the same way we do with the Logback and the Log4j 2 versions?

xbreizh replied:
Done

xbreizh committed:
review changes

xbreizh committed:
Fix rendering

xbreizh committed:
update Readme

xbreizh committed:
cleanup

xbreizh committed:
fix logic
